### PR TITLE
Backend auth #7

### PR DESCRIPTION
**Pull request by grogsy**
PR for #7

### CODE REVIEW

**Comment by powerseed** [comment]
I am not very proficent in Node.js, but I can tell this is high quality code. Good job~

**Review by powerseed — APPROVED** [review]
I am not very proficent in Node.js, but I can tell this is high quality code. Good job~

**Comment by powerseed** [comment]
1. In the User model, do we want to set the password unique?
2. It seems that now after a user registered, we don't make her login automatically. Do we want that?

**Comment by jt5301** [comment]
Regarding your second point, I think I can make that happen on the front end. @grogsy can you add a res.status(200) on successful registration? I could maybe check for that and redirect them to the main component afterwards.

edit : nvm 200 is automatic i think so doesn't matter

**Comment by grogsy** [comment]
@jt5301 I'm working on a change right now to return status codes for error responses so they show up on the response bodies as well.

@powerseed I think its fine if the password field is non-unique. What do you guys think @jt5301 @aeciorc

**Comment by jt5301** [comment]
i think it's fine too.
When a user is newly registered, should they also be given a token when you send the response back? I noticed you were just sending the newly created user. Unless we decide to make them log in manually after signing up?

Also is the idea / is it ok if I just store the token in localstorage on the front?

**Comment by grogsy** [comment]
> When a user is newly registered, should they also be given a token when you send the response back? I noticed you were just sending the newly created user. Unless we decide to make them log in manually after signing up?

That was the impression I was getting: that they should manually log in after. The reasoning was because signing the token is handled by the /login route, whereas I thought /register is solely responsible for storing a new user on the backend. However, I'm up for implementing /register to handle tokens as well, just want to make sure this is what we want.

edit: now that I think about, I think it should fine as is not to have /register handle token logic. Reason being that as @jt5301 mentioned, automatic login after registration can be handled on the front-end. I believe this is possible by leveraging the user object returned on the response body from /register. if its successful it would mean issuing a token isn't necessary because what the token is doing right now is getting that user info. 

> is it ok if I just store the token in localstorage on the front?

I would say yes, but also I don't think thats up to me per se. Although I believe that after logging in the token is attached to a cookie which itself is attached to subsequent requests.

**Comment by powerseed** [comment]
> @jt5301 I'm working on a change right now to return status codes for error responses so they show up on the response bodies as well.
> 
> @powerseed I think its fine if the password field is non-unique. What do you guys think @jt5301 @aeciorc

I am so sry I mean "email" not password... It is a typo.

**Comment by grogsy** [comment]
oh... I thought I made it unique but it turns I had not. Good find @powerseed! I'll be fixing that now

edit: emails were actually quasi-unique in the sense that this code handling registration
```
  const existingUser = await User.findOne({ where: { email: req.body.email } });

  if (existingUser) {
    return res.status(400).json({ email: "That email already exists" });
  }
```

was checking if a db entry containing that email value exists. However I will still be adding the unique constraint because it doesn't hurt to

File: server/models/user.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    allowNull: false,
+    get() {
+      // this prevents password from being serialized on the response output
+      // because the password field is now a getter function
+      // Therefore to access the value stored you must call it as:
+      // someUserModel.password()
+      return () => this.getDataValue("password");
+    },
+  },
+});
```
Comment (powerseed) [inline server/models/user.js:27]: According to the signup page, we may need a "company" field here.

Comment (grogsy) [reply server/models/user.js:27]: Oh right, will be pushing for that change now

File: server/models/user.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+// custom instance methods
+User.prototype.checkPassword = async function (inputPassword) {
+  return bcrypt.compare(inputPassword, this.password());
+};
```
Comment (aeciorc) [inline server/models/user.js:32]: Don't you need to await bcrypt.compare?

File: server/routes/middleware/requiresAuth.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+const jwt = require("jsonwebtoken");
+const { createErrorResponse } = require("./util");
+
+module.exports = async (req, res, next) => {
+  const { token } = req.cookies;
+
+  if (!token) {
+    return createErrorResponse(res, 403, { message: "Forbidden" });
+  } else {
```
Comment (aeciorc) [inline server/routes/middleware/requiresAuth.js]: Return 401 here too

File: server/models/user.js (hunk 2 above)
Comment (aeciorc) [reply server/models/user.js:32]: Oh it doesn't matter nvm

File: server/routes/users.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      email: "That email already exists",
+    });
+  }
+
+  const newUser = await User.create({
+    email: req.body.email,
+    password: req.body.password,
+  });
+
+  res.json(newUser);
```
Comment (aeciorc) [inline server/routes/users.js:43]: Also send the token on a succesful signup

Comment (aeciorc) [reply server/routes/users.js:43]: This way the frontend doesn't need to make one more call to login

File: server/test/loginRoute.spec.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
+const chai = require("chai");
+const chaiHttp = require("chai-http");
+const app = require("../app.js");
+const db = require("../db");
+require("dotenv").config();
+
+chai.should();
+chai.use(chaiHttp);
+
+describe("/POST api/users/login", () => {
```
Comment (aeciorc) [inline server/test/loginRoute.spec.js:10]: Always good to have some tests for auth 👍

File: server/routes/middleware/requiresAuth.js (hunk 3 above)
Comment (grogsy) [reply server/routes/middleware/requiresAuth.js]: should I replace 403 with 401?

Comment (grogsy) [reply server/routes/middleware/requiresAuth.js]: update: i replaced it. I understand why it should be 401 now

**Review by aeciorc — APPROVED** [review]

**Pull request merged**